tests/migrations/test_bd57477f19a2.py

```diff
@@ -0,0 +1,233 @@
+    """
+
+    NUM_USERS = 20
+    NUM_SOURCES = 20
+    NUM_REPLIES = 40
+
+    def __init__(self, homedir):
+        subprocess.check_call(["sqlite3", os.path.join(homedir, "svs.sqlite"), ".databases"])
+        self.session = db.make_session_maker(homedir)()
+
+    def load_data(self):
+        for source_id in range(1, self.NUM_SOURCES + 1):
+            add_source(self.session)
+
+            # Add zero to a few messages from each source, some messages are set to downloaded
+            for _ in range(random.randint(0, 3)):
+                add_message(self.session, source_id)
+
+            # Add zero to a few files from each source, some files are set to downloaded
+            for i in range(random.randint(0, 3)):
+                add_file(self.session, source_id)
+
+        self.session.commit()
+
+        for i in range(self.NUM_USERS):
+            if i == 0:
+                # As of this migration, the server tells the client that the associated journalist
+                # of a reply has been deleted by returning "deleted" as the uuid of the associated
+                # journalist. This gets stored as the jouranlist_id in the replies table.
+                #
+                # Make sure to test this case as well.
+                add_user(self.session, "deleted")
+                source_id = random.randint(1, self.NUM_SOURCES)
+                add_reply(self.session, "deleted", source_id)
+            else:
+                add_user(self.session)
+
+        self.session.commit()
+
+        # Add replies from randomly-selected journalists to a randomly-selected sources
+        for _ in range(1, self.NUM_REPLIES):
+            journalist_id = random.randint(1, self.NUM_USERS)
+            source_id = random.randint(1, self.NUM_SOURCES)
+            add_reply(self.session, journalist_id, source_id)
+
+        self.session.commit()
+
+        # Mark some files, messages, and replies as seen
+        sql = "SELECT * FROM files"
+        files = self.session.execute(text(sql)).fetchall()
+        for file in files:
+            if random.choice([0, 1]):
+                selected_journo_id = random.randint(1, self.NUM_USERS)
+                mark_file_as_seen(self.session, file.id, selected_journo_id)
+
+        sql = "SELECT * FROM messages"
+        messages = self.session.execute(text(sql)).fetchall()
+        for message in messages:
+            if random.choice([0, 1]):
+                selected_journo_id = random.randint(1, self.NUM_USERS)
+                mark_message_as_seen(self.session, message.id, selected_journo_id)
+
+        sql = "SELECT * FROM replies"
+        replies = self.session.execute(text(sql)).fetchall()
+        for reply in replies:
+            if random.choice([0, 1]):
+                selected_journo_id = random.randint(1, self.NUM_USERS)
+                mark_reply_as_seen(self.session, reply.id, selected_journo_id)
+
```

Comment:
would it make sense as part of the testing to check the contents by retrieving files/messages/replies?

Reply:
what would we be testing? here we are populating the seen table with some data so that we can test that the downgrade works when there's data in the tables. maybe it's not really even necessary to do any of this since we already know that dropping a table with data works just fine? open to more ideas!

Reply:
Good point. I agree with you, the suggestion proposed is not necessary here.